ddtrace/api.py

```diff
@@ -14,43 +14,66 @@ class API(object):
     """
     Send data to the trace agent using the HTTP protocol and JSON format
     """
-    def __init__(self, hostname, port, wait_response=False, headers=None, encoder=None):
+    def __init__(self, hostname, port, headers=None, encoder=None):
         self.hostname = hostname
         self.port = port
+        self._traces = '/v0.3/traces'
+        self._services = '/v0.3/services'
+        self._compatibility_mode = False
         self._encoder = encoder or get_encoder()
-        self._wait_response = wait_response
 
         # overwrite the Content-type with the one chosen in the Encoder
         self._headers = headers or {}
         self._headers.update({'Content-Type': self._encoder.content_type})
 
+    def _downgrade(self):
+        """
+        Downgrades the used encoder and API level. This method must fallback to a safe
+        encoder and API, so that it will success despite users' configurations. This action
+        ensures that the compatibility mode is activated so that the downgrade will be
+        executed only once.
+        """
+        self._compatibility_mode = True
+        self._traces = '/v0.2/traces'
+        self._services = '/v0.2/services'
+        self._encoder = JSONEncoder()
+        self._headers.update({'Content-Type': self._encoder.content_type})
+
     def send_traces(self, traces):
         if not traces:
             return
         start = time.time()
         data = self._encoder.encode_traces(traces)
-        response = self._send_span_data(data)
+        response = self._put(self._traces, data)
+
+        # the API endpoint is not available so we should downgrade the connection and re-try the call
+        if response.status in [404, 415] and self._compatibility_mode is False:
```

Comment:
we can make this more generic related to the ``API`` instance state. But I think that this is not the right moment until agent APIs are near to a ``v1.0``